nwb_conversion_tools/datainterfaces/behavior/movie/movie_utils.py

```diff
@@ -13,6 +15,134 @@
 PathType = Union[str, Path]
 
 
+class VideoCaptureContext:
+    def __init__(self, *args, stub=False, **kwargs):
+        self.vc = cv2.VideoCapture(*args, **kwargs)
+        self._args = args
+        self._kwargs = kwargs
+        self.stub = stub
+        self._current_frame = 0
+        self.frame_count = self.get_movie_frame_count()
+        assert self.frame_count > 0, "movie contains no frames"
+        self.fps = self.get_movie_fps()
+        self.frame = self.get_movie_frame(0)
+        assert self.frame is not None, "unable to read the movie file provided"
+
+    def get_movie_timestamps(self):
+        """
+        Return numpy array of the timestamps for a movie file.
+
+        """
+        if not self.vc.isOpened():
+            raise ValueError("movie file is not open")
+        ts = [self.vc.get(cv2.CAP_PROP_POS_MSEC)]
+        for i in tqdm(range(1, self.get_movie_frame_count()), desc="retrieving video timestamps"):
+            self.current_frame = i
+            ts.append(self.vc.get(cv2.CAP_PROP_POS_MSEC))
+        self.current_frame = 0
+        return np.array(ts)
```

Comment:
Do you know if this iteration is done lazily with respect to the reading of video frames? That is, whenever it tries to retrieve the `CAP_PROP_POS_MSEC` for a frame, does it also have to decompress/load the frame data?

Reply:
I think, based on [this](https://stackoverflow.com/questions/47743246/getting-timestamp-of-each-frame-in-a-video), OpenCV assumes the frame rate to be constant and the timestamps retrieved using `cv2.CAP_PROP_POS_MSEC` is generated from the frame_rate. So it might be better that we give this up all together and just build from the frame rate https://github.com/catalystneuro/nwb-conversion-tools/pull/392/commits/4536888fedb0caed486a19ddb8b392491da03d95

Reply:
That discussion indicates specificity to FFMPEG; I recall trying this out on a variety of actual real-world video files and in contrast found them to be highly irregular to the sampling rate

Reply:
There is a huge difference in time when retrieving timestamps for the two cases below:
```python
import cv2
from time import time
from tqdm import tqdm
file_path = "path_to_file.mp4"

# retrieve frames:
vc = cv2.VideoCapture(file_path)
no_frames = int(vc.get(cv2.CAP_PROP_FRAME_COUNT))
fps = vc.get(cv2.CAP_PROP_FPS)
print(f"no frames: {no_frames}, fps: {fps}")


start = time()
ts1 = []
for no in tqdm(range(no_frames//3000), desc="setting frames"):
    vc.set(cv2.CAP_PROP_POS_FRAMES, no)
    ts1.append(vc.get(cv2.CAP_PROP_POS_MSEC))

print(f"using set total time: {time()-start}")
vc.release()


vc = cv2.VideoCapture(file_path)
start = time()
ts2 = []
for no in tqdm(range(no_frames//3000),desc="retrieve frames"):
    success, frame = vc.read()
    ts2.append(vc.get(cv2.CAP_PROP_POS_MSEC))
    if not success:
        break

print(f"using set total time: {time()-start}")
vc.release()
```
Moreover, the ts1 and ts2 values differ. ts2 is accurate (checking against frame rate)
Can you try it on your end for a large video file?
The one I tried was with over 100k frames with 25 fps.

Reply:
I have made the changes here: https://github.com/catalystneuro/nwb-conversion-tools/pull/392/commits/75fbe1adaafddf41a4abd08b60da014722b121c0

Reply:
Good to know these kinds of things; very interesting that there is not only a speed difference but a value difference, which indicates need for some investigation. I'll run some on my end as well, then.

Reply:
OK here's some reading...

Unreliability of FFMPEG documentation, and how opencv relies on it a lot more than we would like underneath: https://stackoverflow.com/questions/70227392/opencvs-cap-setcv2-cap-prop-pos-msec-changes-to-unreachable-timestamp

Official maintainer comments on the unreliability of frame seeking and how frame-specific properties ought to really only be referred to on grabbed frames: https://github.com/opencv/opencv/issues/15749

Very acquisition-specific (non-python) approaches showing irregular timestamps can indeed be encoded, albeit explicitly: https://stackoverflow.com/questions/28086775/can-i-create-a-vfr-video-from-timestamped-images

More background reading on VFR (Variable Frame Rate): http://avisynth.nl/index.php/VFR#encoding_to_vfr_.28mkv.29

Most common use case in neuroscience I think I've seen is encoding of 'dropped' frames (produces irregular timestamps because some of the normally regular timestamps go missing). Also pretty common in general whenever a frame gets corrupted and we need to essentially skip it; I've run into cases of this during some Buzsaki conversion, but it will take me a while to find the exist files again...

Reply:
Now, what I've found from your script on a random `.mpg` file with 9100 frames at 30 fps: `ts2` is indeed accurate and regular w.r.t. the fps. The `ts1` field is equal to `ts2 / (fps * 100)`, indicating a kind of double-scaling bug where the frame number to time conversion is applied twice or something like that.

There is also the massive difference in time to extract these values, which related to the general seeking behavior for these codecs (almost answer two birds in one stone here); the set frame -> get timestamp (supposedly lazy method) takes me around 8 times longer than just reading each frame in succession and throwing it away (the accurate `ts2` method).

So, given these results, for our implementation here I really have to

(a) ask if we need to support allowing users to 'seek' to a particular frame for some reason. It seems very cost ineffective and potentially dangerous depending on how well the user knows these sorts of details that we are exploring

(b) it would be great if we could extract frames to be written to `data` and timestamps simultaneously, but that is only possible when the ultimate write method (outside the scope of this PR) is to not use iterative write (I.e., a data buffer). Otherwise, we have to fully extract the timestamps to determine if they are regular, in order to determine if we want to set the timestamps for the `ImageSeries`... So I guess we really should just implement and use the `ts2` method for all timestamp extraction by default.